// @bscotch/utility/build/lib/wait.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.waits = exports.waitForTick = exports.resolveInNextTick = exports.waitForSeconds = exports.resolveInSeconds = exports.wait = exports.waitForMillis = exports.resolveInMillis = void 0;
/** Get a promise that resolves in some number of milliseconds. */
function resolveInMillis(millis) {
    return new Promise((res) => setTimeout(res, millis));
}
exports.resolveInMillis = resolveInMillis;
/** @alias resolveInMillis */
exports.waitForMillis = resolveInMillis;
/** @alias resolveInMillis */
exports.wait = resolveInMillis;
/** Get a promise that resolves in some number of seconds. */
function resolveInSeconds(seconds) {
    return resolveInMillis(seconds * 1000);
}
exports.resolveInSeconds = resolveInSeconds;
/** @alias resolveInSeconds */
exports.waitForSeconds = resolveInSeconds;
function resolveInNextTick() {
    return new Promise((res) => setImmediate(res));
}
exports.resolveInNextTick = resolveInNextTick;
/** @alias resolveInNextTick */
exports.waitForTick = resolveInNextTick;
exports.waits = {
    resolveInMillis,
    resolveInSeconds,
    resolveInNextTick,
    wait: exports.wait,
    waitForMillis: exports.waitForMillis,
    waitForSeconds: exports.waitForSeconds,
    waitForTick: exports.waitForTick,
};
